File: server/src/sessions-api.ts
import {
  deleteSession,
  forkSession,
  getSessionInfo,
  getSessionMessages,
  listSessions,
  renameSession,
  tagSession,
  type SDKSessionInfo,
  type SessionMessage,
} from '@anthropic-ai/claude-agent-sdk';
import type { FastifyInstance } from 'fastify';

import type { SessionSummary } from '@cc/shared';

import { isPathAllowed } from './config.js';
import { messageToWire } from './serialize.js';

function requirePath(cwd: string | undefined): string {
  if (!cwd) throw new Error('missing cwd');
  if (!isPathAllowed(cwd)) throw new Error(`Path not allowed: ${cwd}`);
  return cwd;
}

function toSummary(s: SDKSessionInfo): SessionSummary {
  return {
    session_id: s.sessionId,
    summary: s.summary ?? s.firstPrompt ?? null,
    title: s.customTitle ?? null,
    tags: s.tag ? [s.tag] : [],
    last_modified: s.lastModified ?? null,
    cwd: s.cwd ?? null,
    num_messages: null,
    total_cost_usd: null,
  };
}

export async function registerSessionsApi(app: FastifyInstance): Promise<void> {
  /**
   * List sessions for a given working directory.
   * The SDK returns all sessions globally (the `dir` filter is loose), so we
   * filter strictly to sessions whose cwd matches the requested path.
   */
  app.get<{
    Querystring: { cwd: string; limit?: string; offset?: string; include_subdirs?: string };
  }>('/sessions', async (req) => {
    const cwd = requirePath(req.query.cwd);
    const limit = req.query.limit ? Number(req.query.limit) : 200;
    const offset = req.query.offset ? Number(req.query.offset) : 0;
    const includeSubdirs = req.query.include_subdirs === 'true';

    const all = await listSessions({ dir: cwd, limit: 1000, offset: 0 });
    const filtered = all.filter((s) => {
      const sCwd = s.cwd ?? '';
      if (!sCwd) return false;
      if (sCwd === cwd) return true;
      if (includeSubdirs && sCwd.startsWith(cwd + '/')) return true;
      return false;
    });
    return filtered.slice(offset, offset + limit).map(toSummary);
  });

  app.get<{ Params: { id: string }; Querystring: { cwd: string } }>(
    '/sessions/:id',
    async (req, reply) => {
      const cwd = requirePath(req.query.cwd);
      const info = await getSessionInfo(req.params.id, { dir: cwd });
      if (!info) {
        reply.code(404);
        return { detail: 'Session not found' };
      }
      return info;
    },
  );

  /**
   * Paginated session messages — reverse-infinite-scroll friendly.
   *
   *   GET /sessions/:id/messages?cwd=...&limit=50
   *     → 最后 50 条（首屏），按 chronological order 升序。
   *
   *   GET /sessions/:id/messages?cwd=...&limit=50&before_uuid=<uuid>
   *     → 找到该 uuid 在完整链中的位置，取它**前面**的 50 条。
   *
   * 响应：
   *   { messages: [...], has_more: boolean, total: number }
   *
   * SDK 的 getSessionMessages(offset/limit) 是从开头算 offset，
   * 而我们需要"最近 N 条"语义；最稳妥的方式是先一次读全（JSONL parse 快、
   * 本地磁盘），再在内存里 slice。1000 条以内毫秒级。
   */
  app.get<{
    Params: { id: string };
    Querystring: { cwd: string; limit?: string; before_uuid?: string };
  }>('/sessions/:id/messages', async (req) => {
    const cwd = requirePath(req.query.cwd);
    const limit = req.query.limit ? Math.max(1, Math.min(500, Number(req.query.limit))) : 50;
    const beforeUuid = req.query.before_uuid;

    const all: SessionMessage[] = await getSessionMessages(req.params.id, { dir: cwd });
    const total = all.length;

    let upper = total; // exclusive
    if (beforeUuid) {
      const idx = all.findIndex((m) => (m as { uuid?: string }).uuid === beforeUuid);
      if (idx > 0) upper = idx;
    }
    const lower = Math.max(0, upper - limit);
    const slice = all.slice(lower, upper);

    return {
      messages: slice.map((sm) => {
        const rawTs = (sm as { timestamp?: string | number }).timestamp;
        const ts =
          typeof rawTs === 'string' ? Date.parse(rawTs) :
          typeof rawTs === 'number' ? rawTs :
          null;
        const wire = messageToWire(sm);
        return {
          uuid: (sm as { uuid?: string }).uuid ?? null,
          parent_uuid: (sm as { parent_uuid?: string }).parent_uuid ?? null,
          timestamp: ts,
          message: wire ? { ...wire, timestamp: ts ?? undefined } : sm,
        };
      }),
      has_more: lower > 0,
      total,
    };
  });

  app.post<{
    Params: { id: string };
    Querystring: { cwd: string; title: string };
  }>('/sessions/:id/rename', async (req) => {
    const cwd = requirePath(req.query.cwd);
    await renameSession(req.params.id, req.query.title, { dir: cwd });
    return { ok: true };
  });

  app.post<{
    Params: { id: string };
    Querystring: { cwd: string; tag: string };
  }>('/sessions/:id/tag', async (req) => {
    const cwd = requirePath(req.query.cwd);
    await tagSession(req.params.id, req.query.tag, { dir: cwd });
    return { ok: true };
  });

  app.post<{
    Params: { id: string };
    Querystring: { cwd: string; title?: string };
  }>('/sessions/:id/fork', async (req) => {
    const cwd = requirePath(req.query.cwd);
    const result = await forkSession(req.params.id, {
      dir: cwd,
      ...(req.query.title ? { title: req.query.title } : {}),
    });
    return result;
  });

  app.delete<{ Params: { id: string }; Querystring: { cwd: string } }>(
    '/sessions/:id',
    async (req) => {
      const cwd = requirePath(req.query.cwd);
      await deleteSession(req.params.id, { dir: cwd });
      return { ok: true };
    },
  );
}

File: app/lib/api/sessions_api.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

class SessionSummary {
  final String sessionId;
  final String? summary;
  final String? title;
  final List<String> tags;
  final int? lastModified;
  final String? cwd;
  final int? numMessages;
  final double? totalCostUsd;

  SessionSummary({
    required this.sessionId,
    this.summary,
    this.title,
    this.tags = const [],
    this.lastModified,
    this.cwd,
    this.numMessages,
    this.totalCostUsd,
  });

  String get displayTitle => title ?? summary ?? '(Untitled)';

  factory SessionSummary.fromJson(Map<String, dynamic> json) {
    final tagsRaw = json['tags'] as List? ?? const [];
    return SessionSummary(
      sessionId: json['session_id'] as String,
      summary: json['summary'] as String?,
      title: json['title'] as String?,
      tags: tagsRaw.map((e) => e.toString()).toList(),
      lastModified: (json['last_modified'] as num?)?.toInt(),
      cwd: json['cwd'] as String?,
      numMessages: (json['num_messages'] as num?)?.toInt(),
      totalCostUsd: (json['total_cost_usd'] as num?)?.toDouble(),
    );
  }
}

class SessionsApi {
  final String baseUrl;
  SessionsApi(this.baseUrl);

  Uri _u(String path, [Map<String, dynamic>? query]) {
    final q = query?.map((k, v) => MapEntry(k, v.toString()));
    return Uri.parse('$baseUrl$path').replace(queryParameters: q);
  }

  Future<List<SessionSummary>> list(String cwd, {int limit = 50}) async {
    final resp = await http.get(_u('/sessions', {'cwd': cwd, 'limit': limit}));
    if (resp.statusCode != 200) {
      throw Exception('list_sessions HTTP ${resp.statusCode}: ${resp.body}');
    }
    final list = jsonDecode(resp.body) as List;
    return list.map((e) => SessionSummary.fromJson(Map<String, dynamic>.from(e))).toList();
  }

  Future<void> rename(String sessionId, String cwd, String title) async {
    final resp = await http.post(_u('/sessions/$sessionId/rename', {'cwd': cwd, 'title': title}));
    if (resp.statusCode != 200) throw Exception(resp.body);
  }

  Future<void> tag(String sessionId, String cwd, String tag) async {
    final resp = await http.post(_u('/sessions/$sessionId/tag', {'cwd': cwd, 'tag': tag}));
    if (resp.statusCode != 200) throw Exception(resp.body);
  }

  Future<String?> fork(String sessionId, String cwd, {String? title}) async {
    final query = {'cwd': cwd};
    if (title != null) query['title'] = title;
    final resp = await http.post(_u('/sessions/$sessionId/fork', query));
    if (resp.statusCode != 200) throw Exception(resp.body);
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    return body['session_id'] as String?;
  }

  Future<void> delete(String sessionId, String cwd) async {
    final resp = await http.delete(_u('/sessions/$sessionId', {'cwd': cwd}));
    if (resp.statusCode != 200) throw Exception(resp.body);
  }
}
